adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,508 @@
+		CFRelease(tlsio_instance->remote_host);
+		CFRelease(tlsio_instance->sockRead);
+		CFRelease(tlsio_instance->sockWrite);
+        free(tlsio_instance);
+    }
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_026: [ The tlsio_appleios_open shall star the process to open the ssl connection with the host provided in the tlsio_appleios_create. ]*/
+int tlsio_appleios_open_async(
+    CONCRETE_IO_HANDLE tlsio_handle,
+    ON_IO_OPEN_COMPLETE on_io_open_complete,
+    void* on_io_open_complete_context,
+    ON_BYTES_RECEIVED on_bytes_received,
+    void* on_bytes_received_context,
+    ON_IO_ERROR on_io_error,
+    void* on_io_error_context)
+{
+    int result = 0;
+    TLS_IO_INSTANCE* tlsio_instance = (TLS_IO_INSTANCE*)tlsio_handle;
+
+    if (tlsio_handle == NULL)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_036: [ If the tlsio_handle is NULL, the tlsio_appleios_open shall not do anything, and return _LINE_. ]*/
+        LogError("NULL TLS handle.");
+        result = __FAILURE__;
+    }
+    else
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_004: [ The tlsio_appleios shall call the callbacks functions defined in the `xio.h`. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_006: [ The tlsio_appleios shall return the status of all async operations using the callbacks. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_007: [ If the callback function is set as NULL. The tlsio_appleios shall not call anything. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_028: [ The tlsio_appleios_open shall store the provided on_io_open_complete callback function address. ]*/
+        tlsio_instance->on_io_open_complete = on_io_open_complete;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_029: [ The tlsio_appleios_open shall store the provided on_io_open_complete_context handle. ]*/
+        tlsio_instance->on_io_open_complete_context = on_io_open_complete_context;
+
+        /* Codes_SRS_TLSIO_APPLEIOS_32_030: [ The tlsio_appleios_open shall store the provided on_bytes_received callback function address. ]*/
+        tlsio_instance->on_bytes_received = on_bytes_received;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_031: [ The tlsio_appleios_open shall store the provided on_bytes_received_context handle. ]*/
+        tlsio_instance->on_bytes_received_context = on_bytes_received_context;
+
+        /* Codes_SRS_TLSIO_APPLEIOS_32_032: [ The tlsio_appleios_open shall store the provided on_io_error callback function address. ]*/
+        tlsio_instance->on_io_error = on_io_error;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_033: [ The tlsio_appleios_open shall store the provided on_io_error_context handle. ]*/
+        tlsio_instance->on_io_error_context = on_io_error_context;
+
+        if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_ERROR) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPEN) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSING))
```

Comment:
>TLSIO_APPLEIOS_STATE_CLOSING [](start = 38, length = 28)

Wrong behavior. Call open on CLOSING is a wrong command, so it must log an error, return __FAILURE__, but **do not** change the state to ERROR, remaining on CLOSING.